### ISSUE TITLE
Timesteps produced by the CTC decoder are wrong.

### ISSUE THREAD

**Issue by godefv** [mozilla/DeepSpeech#3180]
Since commit 33760a6b, the following code was removed from file `native_client/ctcdecode/path_trie.cpp`.

```
      // If existing child matches this new_char but had a lower probability,
      // and it's a leaf, update its timestep to new_timestep.
      // The leak check makes sure we don't update the child to have a later
      // timestep than a grandchild.
      if (child->second->log_prob_c < cur_log_prob_c &&
          child->second->children_.size() == 0) {
        child->second->log_prob_c = cur_log_prob_c;
        child->second->timestep = new_timestep;
      }
```

Therefore, the timesteps are now the first time at which each path was created, which is almost always too early. Often, the peak of probability of a letter is outside the time range given by the decoder.

Anyway, there is a way to get even better timesteps. If you want, I can submit a PR for that.
For example, on a sequence like "ABB" with probabilities 75% for the first B and 80% for the second B, the above code would put the timestep of the letter B on the second B, whereas it should be on the first to include both Bs.

**Comment by reuben** [comment]
> Once upon a time, the ctc decoder tried to produce accurate timesteps.

I don't know if you're just trying to be playful, but it's coming across as rudeness. The implied message is that our code is not even trying to be accurate anymore, as if we're either making it worse on purpose or too dumb to realize what we're doing.

> But since commit 33760a6, the following code was removed

The surrounding discussion is here: https://github.com/mozilla/DeepSpeech/issues/2867#issuecomment-612595999

If you can submit a PR that does not regress that behavior and improves timings, it is more than welcome. I will however repeat what I said in other timing issues, which is that if you want to significantly rework the timing logic, then you need to also present hard data to show what you're talking about. "is almost always too early" is not enough. Provide affected samples, compare before and after, contribute some regression tests for the timing logic.

**Comment by godefv** [comment]
I am sorry, I didn't want to sound rude.

**Comment by godefv** [comment]
Actually, it seems to me that the removal of the timestep update might have been accidental.
Maybe only the probability update was intended ?

**Comment by godefv** [comment]
I have edited my first message to remove the failed funny part.

**Comment by reuben** [comment]
Thanks, I removed the surrounding discussion as well.

**Comment by reuben** [comment]
> Actually, it seems to me that the removal of the timestep update might have been accidental.
> Maybe only removing the probability update was intended ?

I removed the timestep update because I didn't want to "stack heuristics", and I thought the timestep update and the late beam expansion logic (which is currently in the tree) would achieve the same result. I assume you have seen evidence to the contrary, hence this issue.

If adding the timestep update improves the bad timings you're seeing, then maybe it is the right thing to do. Can you share some audio files with timing results before and after the change to exemplify what is going on?

**Comment by godefv** [comment]
I understand better now. The issue is more subtle than I thought.

I have used a slightly different method (comparing probabilities of the complete path, instead of just the last bit), which provides much better accuracy in my case. 
However, I have just realized that this method is not flawless either, so it is also a heuristic.

Here is a small example. In my case, the argmax of the logits gives the following (it is french) :
```
tou_________________________________________________ss  les  _a__mouurreuux  de  se_p_ort__ diiivverrr_   ss'enn__ _r_é___j_uuii__rr__on_t____...
```
The decoded words are "tous les amoureux de sport divers s'en réjouiront".

My first order criterion is that the time range for a word should always fit the argmax output, at least when the confidence of the acoustic model is very high. Namely, "tous" should fit exactly "tou_________________________________________________ss", and "amoureux" should fit either "_a__mouurreuux" or " a__mouurreuux". But the range of "divers" is difficult to get right because it can leak onto the next word, so several possibilities are acceptable. 

With the current code, there is often an offset of about one letter between the expected range and the predicted range. In the above example, the predicted range for "amoureux" corresponds to "' _a__mouurreuu'". I consistenty get better matches by updating the timestep when the new path probability is better than the former.

I will come back on this with more data when I am back from vacation (in 3 weeks). I first thought that it was a small issue with an easy fix, but it is not so simple.

**Comment by godefv** [comment]
I am closing this issue for now. If I gather enough evidence of improvement, I will submit a PR.
Thank you for your explanations.

**Comment by reuben** [comment]
Thanks for the analysis, I understand better what you are seeing now.

In the mean time I also remembered an additional reason why I removed the timestep update. It was because makes it much harder to understand the behavior of the decoder. It modifies the PathTrie in-place, and because PathTrie nodes are shared between different beams, changing the timestep and probability of a node also changes every single beam that includes it. This makes it much harder to understand the implications of the heuristic on the global output due to compounding side effects.

The old code "fixed" this problem in a crude way, by only updating leaf nodes, but that is just a band-aid. The proper implementation would require handling the whole subtree in a way that doesn't break the outputs. I tried a few quick approaches but they were either impractical due to performance or didn't work.

One idea I never tried because it was too much work, was either using an immutable (copy on write) prefix tree data structure or just creating a new PathTrie instance and splicing it into the beam when applying that heuristic. The idea is that it would then keep the timestep updates from breaking unrelated beams that shared that node at the old timestep and must not be changed. Or maybe not, I haven't tried it ;)

**Comment by godefv** [comment]
I have thought about creating a new instance as well. I think it is the right thing to do at least in some cases like path "AB_CD" vs path "A__BD" will give two different global transcriptions "ABCD" and "ABD" where the two versions of the decoded "B" really have different timesteps. But it might not be easy to detect.

There is at least, one case where the update is always good, when comparing two paths with the same length and the same transcription. For example, "ABB" vs "A_B", the timestep of "B" should be from the most probable path.
